### PR TITLE
Payment Blocks: add button typography panel to payment button

### LINKED ISSUE #25928: Payment Button Block: Add Dimensions and Typography Panels
### Impacted plugin

Jetpack

### What

Add the Dimensions and Typography panels to the Payment Button block

### How

This will make the Payment Button block work like the Core Button block.

Screenshot 2022-08-30 at 13 59 15Screenshot 2022-08-30 at 13 59 04

_Note: the images above show the panels with all the optional content enabled._

### PR DESCRIPTION
**Pull request by rcrdortiz**
Fixes #25928

#### Changes proposed in this Pull Request:

* We've added support for the Core typography setting to our Jetpack/Button block.

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Does this pull request change what data or activity we track or use?
No

#### Testing instructions:

* Load the changes on a Simple, Atomic and JN site.
* Open or edit a post.
* Add a recurring payments block.
* In the sidebar you should have the typography controls available (Font controls are theme dependant, 22 theme does have these controls)
* The controls should work as expected.
* Save the post and open the post contents, the typography changes should transfer.

### CODE REVIEW

File: projects/plugins/jetpack/extensions/blocks/button/button.php [hunk 1]
```diff
@@ -105,6 +105,10 @@ function get_button_classes( $attributes ) {
 	$has_named_gradient          = array_key_exists( 'gradient', $attributes );
 	$has_custom_gradient         = array_key_exists( 'customGradient', $attributes );
 	$has_border_radius           = array_key_exists( 'borderRadius', $attributes );
+	$has_font_size               = array_key_exists( 'fontSize', $attributes );
+	$has_typography_styles       = array_key_exists( 'styles', $attributes ) && array_key_exists( 'typography', $attributes['styles'] );
+	$has_custom_font_size        = ! $has_font_size && $has_typography_styles && array_key_exists( 'fontSize', $attributes['styles']['typography'] );
+	$has_custom_font_size        = ! $has_font_size && $has_typography_styles && array_key_exists( 'fontSize', $attributes['styles']['typography'] );
```
Comment (lukaszwilczakwilnet) [inline projects/plugins/jetpack/extensions/blocks/button/button.php]: line 110 and 111 are identical - duplicated.

File: projects/plugins/jetpack/extensions/blocks/button/button.php [hunk 2]
```diff
@@ -105,6 +105,10 @@ function get_button_classes( $attributes ) {
 	$has_named_gradient          = array_key_exists( 'gradient', $attributes );
 	$has_custom_gradient         = array_key_exists( 'customGradient', $attributes );
 	$has_border_radius           = array_key_exists( 'borderRadius', $attributes );
+	$has_font_size               = array_key_exists( 'fontSize', $attributes );
```
Comment (lukaszwilczakwilnet) [inline projects/plugins/jetpack/extensions/blocks/button/button.php:108]: Question: variables $has_font_size, $has_border_radius, $has_font_size is not used yet. Is it some preparation for future work?

Comment (rcrdortiz) [reply projects/plugins/jetpack/extensions/blocks/button/button.php:108]: The PR is WIP

File: projects/plugins/jetpack/extensions/blocks/button/button.php (hunk 1 above)
Comment (rcrdortiz) [reply projects/plugins/jetpack/extensions/blocks/button/button.php]: The PR is WIP

**Comment by rcrdortiz** [comment]
Finished implementing the functionality. This PR can now be reviewed.

File: projects/plugins/jetpack/extensions/blocks/button/edit.js [hunk 3]
```diff
@@ -45,20 +45,25 @@ export function ButtonEdit( props ) {
...
+		{
+			'has-background': backgroundColor.color || gradientValue,
+			[ backgroundColor.class ]: ! gradientValue && backgroundColor.class,
+			'has-text-color': textColor.color,
+			[ textColor.class ]: textColor.class,
+			[ gradientClass ]: gradientClass,
+			'no-border-radius': 0 === borderRadius,
+			'has-custom-width': !! width,
+		},
+		fontSize && { [ `has-${ fontSize }-font-size` ]: true }
```
Comment (mmtr) [inline projects/plugins/jetpack/extensions/blocks/button/edit.js]: I think this works too, so we can keep using just two arguments, right?

```suggestion
		{
			'has-background': backgroundColor.color || gradientValue,
			[ backgroundColor.class ]: ! gradientValue && backgroundColor.class,
			'has-text-color': textColor.color,
			[ textColor.class ]: textColor.class,
			[ gradientClass ]: gradientClass,
			'no-border-radius': 0 === borderRadius,
			'has-custom-width': !! width,
			[ `has-${ fontSize }-font-size` ]: !! fontSize,
		},
```

File: projects/plugins/jetpack/extensions/blocks/button/edit.js [hunk 4]
```diff
@@ -45,20 +45,25 @@ export function ButtonEdit( props ) {
...
+	);
 
 	const buttonStyles = {
 		...( ! backgroundColor.color && gradientValue
 			? { background: gradientValue }
 			: { backgroundColor: backgroundColor.color } ),
+		fontSize: attributes.style?.typography?.fontSize,
```
Comment (mmtr) [inline projects/plugins/jetpack/extensions/blocks/button/edit.js:63]: I don't understand this. In the line 27 above we're getting the font size from `attributes.fontSize` and here we're getting it from `attributes.style.typography.fontSize`. Do we have two different attributes?

**Review by mmtr — COMMENTED** [review]
This indeed adds the typography controls to the `jetpack/button` block, but I think https://github.com/Automattic/jetpack/issues/25928 asked to add them to the `jetpack/recurring-payments` block, and I'm still unable to change the typography when I select the parent Payment button block:

Screen Shot 2022-09-14 at 11 08 19

I think that ideally we should allow the Payment button block to control the typography of the inner Button block, similarly to the behavior of the width settings.

This is a good first step though and my suggestion shouldn't be considered a blocker. But let's keep https://github.com/Automattic/jetpack/issues/25928 open until the typography can be changed from the parent Payment button block.

File: projects/plugins/jetpack/extensions/blocks/button/edit.js (hunk 4 above)
Comment (rcrdortiz) [reply projects/plugins/jetpack/extensions/blocks/button/edit.js:63]: It depends on how you configure the font size. You can either choose from a small, medium, large and extra-large list or manually choose a specific font size. Each of these methods stores the font size in a different location.
When we use the list we must add a button class, when we use the input field we add a style.

**Comment by rcrdortiz** [comment]
> This indeed adds the typography controls to the `jetpack/button` block, but I think #25928 asked to add them to the `jetpack/recurring-payments` block, and I'm still unable to change the typography when I select the parent Payment button block:
> 
> Screen Shot 2022-09-14 at 11 08 19
> 
> I think that ideally we should allow the Payment button block to control the typography of the inner Button block, similarly to the behavior of the width settings.
> 
> This is a good first step though and my suggestion shouldn't be considered a blocker. But let's keep #25928 open until the typography can be changed from the parent Payment button block.

I'll follow up this PR with another one that can handle the underlying button typography from the recurring-payments block's controls.

File: projects/plugins/jetpack/extensions/blocks/button/edit.js (hunk 3 above)
Comment (rcrdortiz) [reply projects/plugins/jetpack/extensions/blocks/button/edit.js]: Applied

File: projects/plugins/jetpack/extensions/blocks/button/save.js [hunk 5]
```diff
@@ -34,20 +35,25 @@ export default function ButtonSave( { attributes, blockName, uniqueId } ) {
...
+		{
+			'has-text-color': textColor || customTextColor,
+			[ textClass ]: textClass,
+			'has-background': backgroundColor || gradient || customBackgroundColor || customGradient,
+			[ backgroundClass ]: backgroundClass,
+			[ gradientClass ]: gradientClass,
+			'no-border-radius': 0 === borderRadius,
+			'has-custom-width': !! width,
+		},
+		fontSize && { [ `has-${ fontSize }-font-size` ]: true }
```
Comment (mmtr) [inline projects/plugins/jetpack/extensions/blocks/button/save.js]: Same as https://github.com/Automattic/jetpack/pull/26108#discussion_r970549610, this one can be simplified a little bit as well

File: projects/plugins/jetpack/extensions/blocks/button/edit.js (hunk 3 above)
Comment (mmtr) [reply projects/plugins/jetpack/extensions/blocks/button/edit.js]: Looks like there some ESLint errors now, sorry about that!

Comment (rcrdortiz) [reply projects/plugins/jetpack/extensions/blocks/button/edit.js]: Fixed the linting errors.

File: projects/plugins/jetpack/extensions/blocks/button/save.js (hunk 5 above)
Comment (rcrdortiz) [reply projects/plugins/jetpack/extensions/blocks/button/save.js]: Simplified using the same approach as the one suggested.

**Review by mmtr — APPROVED** [review]
Lovely! It works great :shipit:

**Comment by rcrdortiz** [comment]
Deployed: r252643-wpcom

**Pull request merged**